### PR TITLE
Gravity postprocessor

### PR DESCRIPTION
**Pull request by LudovicJnnt**

### CODE REVIEW

File: include/aspect/postprocess/gravity_calculation.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
+  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
```
Comment (naliboff) [inline include/aspect/postprocess/gravity_calculation.h]: Update copyright to 2011 - 2018

File: include/aspect/postprocess/gravity_calculation.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#include <aspect/simulator_access.h>
+
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+
+    /**
+     * A postprocessor that outputs the gravity to file.
```
Comment (naliboff) [inline include/aspect/postprocess/gravity_calculation.h]: Add a more detailed explanation here ... ex: This is a post processor that output the gravity at a point (e.g., satellite) above the model surface for a user-defined range of latitude, longitudes and heights.  The documentation should actually be the same as at the end of the .cc file.

File: include/aspect/postprocess/gravity_calculation.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+      private:
+        /**
+         * Gravity is calculated on a spherical surface at which points are spaced
+         * according to the spherical_spacing values.
+         */
+        double spherical_spacing;
+
+        /**
+         * Gravity is calculated at a certain and constant height.
```
Comment (naliboff) [inline include/aspect/postprocess/gravity_calculation.h]: Alternate description: `Height above the model surface (e.g., 'satellite height') at which the gravity is calculated`

File: include/aspect/postprocess/gravity_calculation.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+         */
+        double satellite_height;
+
+        /**
+         * Gravity difference calculation - theoric. Theoric gravity requires a 
+         * reference density of the all domain.
```
Comment (naliboff) [inline include/aspect/postprocess/gravity_calculation.h]: Change to `The reference density used in the gravity calculation.`

File: include/aspect/postprocess/gravity_calculation.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+         */
+        double reference_density;
+
+        std::vector<double> long_min_max;
```
Comment (naliboff) [inline include/aspect/postprocess/gravity_calculation.h]: Consider splitting this variable into two different variables: `longitude_minimum` and `longitude_maximum`.

Comment (naliboff) [reply include/aspect/postprocess/gravity_calculation.h]: Also add a description of each variable

File: include/aspect/postprocess/gravity_calculation.h [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        std::vector<double> lat_min_max;
```
Comment (naliboff) [inline include/aspect/postprocess/gravity_calculation.h]: Same as above.

File: tests/gravity_point.prm [hunk 7]
```diff
@@ -0,0 +1,114 @@
+# A simple setup for gravity acceleration calculated at a point P
+# manual for more information.
```
Comment (naliboff) [inline tests/gravity_point.prm]: Please include more detailed comments here. Say what post processor you are testing, what the output is supposed to look like (analytical solution?) and what specific features of the gravity post processor you are testing.

File: tests/gravity_point.prm [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  end
+end
+
+
+
+subsection Initial temperature model
+  set Model name = S40RTS perturbation
+
+  subsection S40RTS perturbation
+    #set Function expression = 0
```
Comment (naliboff) [inline tests/gravity_point.prm]: remove this line.

File: tests/gravity_point.prm [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    #set Data directory                    = ~/home/ludovic/aspect/aspect_2018-05-07/aspect/data/initial_temperature/S40RTS/
```
Comment (naliboff) [inline tests/gravity_point.prm]: remove this line.

File: tests/gravity_point.prm [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+
+subsection Mesh refinement
+  set Initial global refinement          = 1
+end
+
+
+
+subsection Postprocess
+  set List of postprocessors = visualization, velocity statistics, temperature statistics, heat flux statistics, gravity calculation
```
Comment (naliboff) [inline tests/gravity_point.prm]: Remove everything hear except `gravity calculation`

File: tests/gravity_point.prm [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+  subsection Visualization
```
Comment (naliboff) [inline tests/gravity_point.prm]: Remove this subsection

File: tests/gravity_point.prm [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    set Number of grouped files       = 1
+  end
+ 
+  subsection Gravity calculation
+    set Spherical spacing             = 180
+    set Satellite height              = 6500000
+    set Reference density             = 3300
+  end
+
+  subsection Depth average
```
Comment (naliboff) [inline tests/gravity_point.prm]: Remove this subsection

File: tests/gravity_point.prm [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    set Time between graphical output = 2
+    set Output format                 = vtu
+  end
+end
+
+
+
+subsection Checkpointing
```
Comment (naliboff) [inline tests/gravity_point.prm]: Remove this subsection

**Review by naliboff — COMMENTED** [review]
General comment:
The .h and .cc files should have the same name. They should also be named such that it is clear exactly what the post processor calculates. Example: satellite_gravity.

File: source/postprocess/satellite_gravity.cc [hunk 14]
```diff
@@ -0,0 +1,362 @@
...
+      FEValues<dim> fe_values (this->get_mapping(),
+                               this->get_fe(),
+                               quadrature_formula,
+                               update_values   |
+                               update_gradients |
+                               update_quadrature_points |
+                               update_JxW_values);
+
+      // Get the value of the outer radius and inner radius
+      const double Ro = dynamic_cast<const GeometryModel::SphericalShell<dim>&>
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: consider renaming to `model_outer_radius` or `geometry_outer_radius`

File: source/postprocess/satellite_gravity.cc [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+                                (this->get_geometry_model()).outer_radius();
+      const double Ri = dynamic_cast<const GeometryModel::SphericalShell<dim>&>
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Same as above

File: source/postprocess/satellite_gravity.cc [hunk 16]
```diff
@@ -0,0 +1,362 @@
...
+                                (this->get_geometry_model()).inner_radius();
+
+      // Get the value of the universal gravitational constant
+      const double G = aspect::constants::big_g;
+
+      std_cxx11::array<double,dim> ecoord;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Rename to something more explicit (e.g., `satellite_coordinates`, etc)

File: source/postprocess/satellite_gravity.cc [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+
+      int hh=1; int ii=1; int jj=1;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Consider renaming variables to something more explicit if it makes sense.

File: source/postprocess/satellite_gravity.cc [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+      double height=0;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Is this the height in one of the loops?

File: source/postprocess/satellite_gravity.cc [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+      double h_spacing=0;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: rename to `height_spacing`

File: source/postprocess/satellite_gravity.cc [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+      double longitude=0;
+      double latitude=0;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Add a comment that this is initial latitude and longitude in the loops.

File: source/postprocess/satellite_gravity.cc [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+      int c = 0;
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+      for (; cell!=endc; ++cell)
+        if (cell->is_locally_owned())
+            c += 1 ;
+
+      // Storage of density values in a vector density_all per mpi 
+      int cell_n_q = c * n_q_points;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: `cell_n_q` is the total number of quadrature points locally owned by the MPI process?

File: source/postprocess/satellite_gravity.cc [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+      std::vector<double> density_all;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: We will want to rename this eventually.

File: source/postprocess/satellite_gravity.cc [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+            this->get_material_model().evaluate(in, out);
+            for (unsigned int q = 0; q < n_q_points; ++q)
+              {
+               density_all[c*n_q_points+q]=out.densities[q];
+              }
+            c += 1;
+          }
+      f2 << profile_satellite_height << ' ' << mapping_satellite_height << '\n';
+
+      // Decide if gravity depth-profile or mapping is activated
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Indicate that `mapping` is using a constant height but varying the latitude/longitude

Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: This 'decision' should be done with a variable like `Satellite gravity analysis type` and the options are `height profile` or `mapping profile`.

File: source/postprocess/satellite_gravity.cc [hunk 24]
```diff
@@ -0,0 +1,362 @@
...
+      if (profile_satellite_height > 0) // gravity depth-profile activated
+        { 
+          longitude = profile_longitude;
+          latitude  = profile_latitude;
+          h_spacing = profile_height_spacing;
+          hh = profile_satellite_height/profile_height_spacing;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: Consider renaming to something like `number_satellite_height_profile_points`

Comment (naliboff) [reply source/postprocess/satellite_gravity.cc]: You could indicated what this variable is in a comment.

File: source/postprocess/satellite_gravity.cc [hunk 25]
```diff
@@ -0,0 +1,362 @@
...
+        }
+       else
+        {
+          f2 << '\n' << '\n';
+        }       
+ 
+      // loop on r - satellite position [r, ,  ]
+      for (int h=0; h<hh; h++)
+        {
+          ecoord[0]=height+h*h_spacing;
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: This is a bit confusing to read w/out context: Add in a comment that specifies `height` is 0 for the depth profile case or in the mapping case `h` is zero and `hh` is 1.

File: source/postprocess/satellite_gravity.cc [hunk 26]
```diff
@@ -0,0 +1,362 @@
...
+                  double local_gx = 0;
+                  double local_gy = 0;
+                  double local_gz = 0;
+                  double gnorm = 0;
+                  double gtheory = 0;      
+                  double g_diff = 0;      
+                  double local_U = 0;
+                  c = 0;
+              
+                  // For each point (i.e. satellite), the forth integral goes over cells and quadrature
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: fourth

File: source/postprocess/satellite_gravity.cc [hunk 27]
```diff
@@ -0,0 +1,362 @@
...
+
+      // Storage of density values in a vector density_all per mpi 
+      int cell_n_q = c * n_q_points;
+      std::vector<double> density_all;
+      density_all.reserve(cell_n_q); 
+
+      c = 0;
+      cell = this->get_dof_handler().begin_active();
+      for (; cell!=endc; ++cell)
+        if (cell->is_locally_owned())
```
Comment (naliboff) [inline source/postprocess/satellite_gravity.cc]: add"{}" to for loop

**Comment by naliboff** [comment]
@LudovicJnnt - It looks like you may need to rebase so that only the relevant commits show. Let me know if you would like some detailed instructions on how to do this!

**Comment by LudovicJnnt** [comment]
@naliboff - I think I know how to rebase and see what the problem seems to be. However, this is the old branch, I am not even sure there is any relevant commit in there because the names of the .cc and .h here do not correspond in any case to the already merged gravity post-processor (called gravity_point_values.cc/.h). Should we removed this branch then? or close it somehow?

**Comment by naliboff** [comment]
@LudovicJnnt - To be sure I understand, some of the commits here contain features you would like to add to the Gravity Postprocessor, which has already been merged? If so, you could rebase to the master version and then remove any of the new commits that are not relevant. However, perhaps easier to just close this PR and then make a new branch (+ PR) that contains the changes you would like to make.

**Comment by LudovicJnnt** [comment]
@naliboff - yes I close this PR as it is not relevant anymore. There is nothing new (no new features) to be merged. This was just a preliminary PR for us to work on.

**Pull request closed without merging**